File: growing-shop-java/src/main/java/com/example/growingshop/domain/auth/service/RoleService.java
package com.example.growingshop.domain.auth.service;

import com.example.growingshop.domain.auth.domain.Policies;
import com.example.growingshop.domain.auth.domain.Role;
import com.example.growingshop.domain.auth.domain.Roles;
import com.example.growingshop.domain.auth.dto.RoleRequest;
import com.example.growingshop.domain.auth.repository.PolicyRepository;
import com.example.growingshop.domain.auth.repository.RoleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class RoleService {
    private final RoleRepository roleRepository;
    private final PolicyRepository policyRepository;

    public Roles findAll() {
        return new Roles(roleRepository.findAll());
    }

    public Role getById(Long id) {
        return roleRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException(id + " 에 해당하는 role 을 찾을 수 없습니다."));
    }

    public Role findByName(String name) {
        return roleRepository.getByName(name)
                .orElseThrow(() -> new IllegalArgumentException(name + " 에 해당하는 role 을 찾을 수 없습니다."));
    }

    @Transactional
    public Role create(RoleRequest.CreateRole req) {
        Policies policies = new Policies(policyRepository.findAllById(req.getPolicies()));

        return roleRepository.save(req.toEntity(policies));
    }

    @Transactional
    public void changePolicies(RoleRequest.ChangeRolePolicies req) {
        Role role = getById(req.getRole());
        Policies policies = new Policies(policyRepository.findAllById(req.getPolicies()));

        role.changePolicies(policies);
    }
}

File: growing-shop-api/auth/src/main/kotlin/com/example/growingshopauth/auth/service/RoleService.kt
package com.example.growingshopauth.auth.service

import com.example.domain.auth.Policies
import com.example.domain.auth.Role
import com.example.domain.auth.Roles
import com.example.domain.user.UserType
import com.example.growingshopauth.auth.annotation.AccessibleUserTypes
import com.example.growingshopauth.auth.domain.changePolicies
import com.example.growingshopauth.auth.dto.RoleRequest
import com.example.growingshopauth.auth.repository.ExpandRoleRepository
import com.example.repository.auth.PolicyRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class RoleService(
    private val roleRepository: ExpandRoleRepository,
    private val policyRepository: PolicyRepository
) {

    @AccessibleUserTypes(UserType.ADMIN)
    fun findAll(): Roles = Roles(roleRepository.findAll())

    @AccessibleUserTypes(UserType.ADMIN)
    fun getById(id: Long): Role = roleRepository.findById(id)
        .orElseThrow { IllegalArgumentException("$id 에 해당하는 role 을 찾을 수 없습니다.") }

    @AccessibleUserTypes(UserType.ADMIN)
    fun findByName(name: String): Role = roleRepository.getByName(name)
        .orElseThrow { IllegalArgumentException("$name 에 해당하는 role 을 찾을 수 없습니다.") }

    @Transactional
    @AccessibleUserTypes(UserType.ADMIN)
    fun create(req: RoleRequest.CreateRole): Role = roleRepository.save(
        req.toEntity(
            Policies(policyRepository.findAllById(req.policies))
        )
    )

    @Transactional
    @AccessibleUserTypes(UserType.ADMIN)
    fun changePolicies(req: RoleRequest.ChangeRolePolicies) {
        val role = getById(req.role)
        val policies = Policies(policyRepository.findAllById(req.policies))

        role.changePolicies(policies)
    }

    @Transactional
    @AccessibleUserTypes(UserType.ADMIN)
    fun delete(id: Long) {
        roleRepository.deleteById(id)
    }
}
